phoenix-core/src/main/java/org/apache/phoenix/mapreduce/index/IndexUpgradeTool.java

```diff
@@ -0,0 +1,562 @@
+        String schema = SchemaUtil.getSchemaNameFromFullName(dataTable);
+        String table = SchemaUtil.getTableNameFromFullName(dataTable);
+        for(String index: rebuildMap.get(dataTable)) {
+            String outFile = "/tmp/index_rebuild_" + index + "_" + UUID.randomUUID().toString();
+            String indexName = SchemaUtil.getTableNameFromFullName(index);
+            String[]
+                    args =
+                    { "-s", schema, "-dt", table, "-it", indexName, "-direct", "-op", outFile };
+            indexingTool.setConf(conf);
+            if (syncRebuild) {
+                List<String> list = Arrays.asList(args);
+                list.add("-runfg");
+                args = list.toArray(new String[list.size()]);
+            }
+            try {
+                LOGGER.info("Rebuilding index " + index);
+                if (!dryRun) {
+                    indexingTool.run(args);
+                }
+            } catch (Exception e) {
+                LOGGER.severe("Something went wrong while building the index " + index + " " + e);
+                return -1;
+            }
+        }
+        return 0;
+    }
+
+    private boolean extractTablesAndIndexes(PhoenixConnection conn) {
+        String [] tables = inputTables.trim().split(",");
+        PTable dataTable = null;
+        try {
+            for (String tableName : tables) {
+                HashSet<String> physicalIndexes = new HashSet<>();
+                dataTable = PhoenixRuntime.getTableNoCache(conn, tableName);
+                String physicalTableName = dataTable.getPhysicalName().getString();
+                if (!dataTable.isTransactional() && dataTable.getType().equals(PTableType.TABLE)) {
+                    for (PTable indexTable : dataTable.getIndexes()) {
+                        if (indexTable.getIndexType().equals(PTable.IndexType.GLOBAL)) {
+                            physicalIndexes.add(indexTable.getPhysicalName().getString());
+                        }
+                   }
+                    HashSet<String> rebuildIndexes = new HashSet<>();
+                    rebuildIndexes.addAll(physicalIndexes);
+                    if (MetaDataUtil.hasViewIndexTable(conn, dataTable.getPhysicalName())) {
+                        String viewIndexPhysicalName = MetaDataUtil.getViewIndexPhysicalName(physicalTableName);
+                        physicalIndexes.add(viewIndexPhysicalName);
+
+                        ResultSet
+                                rs =
+                                conn.createStatement().executeQuery(
+                                "SELECT DISTINCT TABLE_NAME FROM " + "SYSTEM.CATALOG WHERE COLUMN_FAMILY = \'"
+                                    + viewIndexPhysicalName +"\' AND TABLE_TYPE = \'i\'");
+                        while (rs.next()) {
+                            String viewIndexName = rs.getString(1);
+                            rebuildIndexes.add(viewIndexName);
```

Comment:
For tenant-owned indexes you'll need both the index name (which you have) _and_ the tenant id. Two tenants can create indexes with the same name, so to rebuild them you need to provide the tenant id that created them and now owns them.